renderdoc/driver/ihv/amd/amd_counters.cpp

```diff
@@ -227,140 +368,378 @@ CounterDescription AMDCounters::InternalGetCounterDescription(uint32_t internalI
 
 void AMDCounters::EnableCounter(GPUCounter counter)
 {
-  const uint32_t internalIndex = m_Counters[GPUCounterToCounterIndex(counter)].internalIndex;
+  const uint32_t internalIndex = m_PublicToInternalCounter[counter];
 
-  m_pGPUPerfAPI->GPA_EnableCounter(internalIndex);
+  GPA_Status status = m_pGPUPerfAPI->GPA_EnableCounter(internalIndex);
+  if(AMD_FAILED(status))
+  {
+    GPA_LoggingCallback(GPA_LOGGING_ERROR, FormatErrMessage("Enable counter.", status).c_str());
+  }
 }
 
 void AMDCounters::EnableAllCounters()
 {
-  m_pGPUPerfAPI->GPA_EnableAllCounters();
+  GPA_Status status = m_pGPUPerfAPI->GPA_EnableAllCounters();
+  if(AMD_FAILED(status))
+  {
+    GPA_LoggingCallback(GPA_LOGGING_ERROR, FormatErrMessage("Enable all counters.", status).c_str());
+  }
 }
 
 void AMDCounters::DisableAllCounters()
 {
-  m_pGPUPerfAPI->GPA_DisableAllCounters();
+  GPA_Status status = m_pGPUPerfAPI->GPA_DisableAllCounters();
+  if(AMD_FAILED(status))
+  {
+    GPA_LoggingCallback(GPA_LOGGING_ERROR, FormatErrMessage("Disable all counters.", status).c_str());
+  }
 }
 
 uint32_t AMDCounters::GetPassCount()
 {
-  gpa_uint32 numRequiredPasses;
-  m_pGPUPerfAPI->GPA_GetPassCount(&numRequiredPasses);
+  gpa_uint32 numRequiredPasses = 0;
+  GPA_Status status = m_pGPUPerfAPI->GPA_GetPassCount(&numRequiredPasses);
+  if(AMD_FAILED(status))
+  {
+    GPA_LoggingCallback(GPA_LOGGING_ERROR, FormatErrMessage("Get pass count.", status).c_str());
+  }
 
   return (uint32_t)numRequiredPasses;
 }
 
 uint32_t AMDCounters::BeginSession()
 {
-  gpa_uint32 sessionID;
+  gpa_uint32 sessionID = 0;
 
-  m_pGPUPerfAPI->GPA_BeginSession(&sessionID);
+  GPA_Status status = m_pGPUPerfAPI->GPA_BeginSession(&sessionID);
+  if(AMD_FAILED(status))
+  {
+    GPA_LoggingCallback(GPA_LOGGING_ERROR, FormatErrMessage("Begin session.", status).c_str());
+  }
 
   return (uint32_t)sessionID;
 }
 
 void AMDCounters::EndSesssion()
 {
-  m_pGPUPerfAPI->GPA_EndSession();
+  GPA_Status status = m_pGPUPerfAPI->GPA_EndSession();
+  if(AMD_FAILED(status))
+  {
+    GPA_LoggingCallback(GPA_LOGGING_ERROR, FormatErrMessage("End session.", status).c_str());
+  }
+}
+
+std::vector<CounterResult> AMDCounters::GetCounterData(uint32_t sessionID, uint32_t maxSampleIndex,
+                                                       const std::vector<uint32_t> &eventIDs,
+                                                       const std::vector<GPUCounter> &counters)
+{
+  std::vector<CounterResult> ret;
+
+  bool isReady = false;
+
+  const uint32_t timeout = 10000;    // ms
+
+  PerformanceTimer startTime;
+
+  do
+  {
+    isReady = IsSessionReady(sessionID);
+    if(!isReady)
+    {
+      std::this_thread::sleep_for(std::chrono::milliseconds(0));
+
+      PerformanceTimer endTime;
+
+      double delta = endTime.GetMilliseconds() - startTime.GetMilliseconds();
```

Comment:
This won't work correctly. The `GetMilliseconds()` returns the time that's passed since the timer started. You should just need a single PerformanceTimer outside the loop and check the `GetMilliseconds()` on it each time against the timeout.